### ISSUE TITLE
[python-package] LightGBMRanker doesn't support the Dataset object?

### ISSUE THREAD

**Issue by dpalbrecht** [microsoft/LightGBM#5732]
I'm not sure whether this is a bug or a feature request, so sorry in advance! Perhaps it's just not documented and I can't find anyone else having tried this or questions regarding it.

I'm trying to train a LightGBMRanker with a LightGBM Dataset object, but the docs say this model requires both `X` and `y` which means I have to set `X=dataset.get_data()` and `y=dataset.get_label()`. This effectively means I'm not using the dataset class. So even if I do this, I can't pass in an `eval_set` that has access to all the original attributes I need for a custom evaluation metric. Is this just not supported at this time?

**Comment by jameslamb** [comment]
Thanks for using LightGBM. In the future, when you open an issue here, please provide the information we've asked for (how you're using LightGBM, what version, etc.).

----------

I'm *guessing* based on what you've asked that you're specifically talking about the Python package, and are wondering if it's possible to pass a `lightgbm.Dataset` object to `lightgbm.sklearn.LGBMRanker`. It is not.

The estimators in `lightgbm.sklearn` (`LGBMClassifier`, `LGBMRegressor`, `LGBMRanker`) take raw data in the form of `numpy` arrays, `pandas` DataFrames, `scipy` matrices, and a few other formats, as described at https://lightgbm.readthedocs.io/en/v3.3.5/pythonapi/lightgbm.LGBMRanker.html.

If you'd like to instead use `lightgbm.Dataset` objects for training data and eval sets, I recommend instead using `lightgbm.train()`. See https://lightgbm.readthedocs.io/en/v3.3.5/pythonapi/lightgbm.train.html#lightgbm.train.

Both methods support custom evaluation metrics, though with slightly different formats for how the evaluation functions are written. See those docs I linked for an explanation of the differences, and the examples in https://github.com/microsoft/LightGBM/tree/master/examples/python-guide for code samples.

----------

> I'm not sure whether this is a bug or a feature request

This is not a bug, and I'm -1 on supporting it as a feature request.

The goal of the `lightgbm.sklearn` interface is to provide compatibility with other tools that can accept scikit-learn compatible estimators (see ["developing scikit-learn estimators"](https://scikit-learn.org/stable/developers/develop.html)). That includes `scikit-learn`'s own model selection and a wide range of other tools. Since those tools all expect `numpy`, `scipy`, and `pandas` data, I think `lightgbm.sklearn`'s interfaces should be limited to accepting those too.

**Comment by dpalbrecht** [comment]
Definitely - my fault for not providing all the requested details. You understood my question. That makes perfect sense now. I'll use `lightgbm.train()` for my current task.
